src/SSLSocket.c

```diff
@@ -608,27 +611,101 @@ int SSLSocket_createContext(networkHandles* net, MQTTClient_SSLOptions* opts)
 			goto free_ctx; /*If we can't load the certificate (chain) file then loading the privatekey won't work either as it needs a matching cert already loaded */
 		}
 
-		if (opts->privateKey == NULL)
-			opts->privateKey = opts->keyStore;   /* the privateKey can be included in the keyStore */
-
-		if (opts->privateKeyPassword != NULL)
-		{
-			SSL_CTX_set_default_passwd_cb(net->ctx, pem_passwd_cb);
-			SSL_CTX_set_default_passwd_cb_userdata(net->ctx, (void*)opts->privateKeyPassword);
-		}
-
-		/* support for ASN.1 == DER format? DER can contain only one certificate? */
-		rc = SSL_CTX_use_PrivateKey_file(net->ctx, opts->privateKey, SSL_FILETYPE_PEM);
-		if (opts->privateKey == opts->keyStore)
-			opts->privateKey = NULL;
-		if (rc != 1)
-		{
-			if (opts->struct_version >= 3)
-				SSLSocket_error("SSL_CTX_use_PrivateKey_file", NULL, net->socket, rc, opts->ssl_error_cb, opts->ssl_error_context);
-			else
-				SSLSocket_error("SSL_CTX_use_PrivateKey_file", NULL, net->socket, rc, NULL, NULL);
-			goto free_ctx;
-		}
+        if ( opts->keyType == NULL
+            || !strncmp(opts->keyType, "PEM", strlen("PEM")))
+        {
+            if (opts->privateKey == NULL)
+                opts->privateKey = opts->keyStore;   /* the privateKey can be included in the keyStore */
+
+            if (opts->privateKeyPassword != NULL)
+            {
+                SSL_CTX_set_default_passwd_cb(net->ctx, pem_passwd_cb);
+                SSL_CTX_set_default_passwd_cb_userdata(net->ctx, (void*)opts->privateKeyPassword);
+            }
+            /* support for ASN.1 == DER format? DER can contain only one certificate? */
+            rc = SSL_CTX_use_PrivateKey_file(net->ctx, opts->privateKey, SSL_FILETYPE_PEM);
+            if (opts->privateKey == opts->keyStore)
+                opts->privateKey = NULL;
+            if (rc != 1)
+            {
+                if (opts->struct_version >= 3)
+                    SSLSocket_error("SSL_CTX_use_PrivateKey_file", NULL, net->socket, rc, opts->ssl_error_cb, opts->ssl_error_context);
+                else
+                    SSLSocket_error("SSL_CTX_use_PrivateKey_file", NULL, net->socket, rc, NULL, NULL);
+                goto free_ctx;
+            }
+        } 
+        else if (!strncmp(opts->keyType, "ENG", strlen("ENG")))
+        {
+            ENGINE *eng = NULL;
+            EVP_PKEY *pkey = NULL;
+
+            if (opts->engineConfFile
+                && opts->engineId
+                && opts->privateKey)
+            {
+                if (!load_openssl_config) 
+                {
+                    if (CONF_modules_load_file(opts->engineConfFile, "openssl_conf", 0) != 1) {
```

Comment:
Why is loading configuration necessary? I would prefer that the default OpenSSL configuration is used instead of allowing applications modify it.
(E.g. https://github.com/dotnet/runtime/pull/88656/files#diff-b4a7a922ce304a412632e8b5a25c5f5a87fe5b3e63f639a5eb63a9c7b02668f9R279)

Either way, the code should allow for the case where `engineConfFile` is NULL.